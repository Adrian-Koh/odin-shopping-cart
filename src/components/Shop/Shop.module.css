.cardImage {
    height: 100px;
    width: auto;
}

main {
    display: grid;
    grid-template-columns: repeat(4, 250px);
    grid-auto-rows: 250px;
    justify-content: center;
    gap: 20px;
}

.card {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: space-between;

    border-radius: 5px;
    box-shadow: 2px 2px 3px 3px rgb(107, 107, 107);
    padding: 5px;
    position: relative;
    background-color: white;
}

h1 {
    margin: 20px;
    text-align: center;
    background-color: white;
    border-radius: 5px;
    padding: 10px;
}

li {
    list-style-type: none;
    margin-bottom: 20px;
    font-size: 16px;
}

ul {
    padding: 0;
}

.container {
    display: grid;
    grid-template-columns: 1fr 5fr;
    justify-items: center;
    width: 95vw;
}

aside {
    margin: 20px;
    padding: 20px;
    border-radius: 10px;
    height: min-content;
    background-color: white;
}

.cardTitle {
    font-weight: 500;
    font-size: 16px;
    text-align: center;
    margin: 5px;
    padding: 0 15px;
}

.price {
    margin: 3px;
    position: absolute;
    top: 3px;
    right: 3px;
    padding: 5px;
    border-radius: 5px;
    background-color: white;
    color: rgb(10, 161, 10);
    font-weight: 700;
    font-size: 16px;
}

nav img {
    height: 30px;
}

.shopNavBar {
    width: fit-content;
    display: flex;
    position: absolute;
    right: 20px;
    top: 20px;
}

.title {
    display: flex;
    width: calc(100vw - 50px);
    justify-content: center;
    position: relative;
    margin-bottom: 20px;
}

.showIcon {
    display: block;
}

.hideIcon {
    display: none;
}

.infoIcon {
    height: 25px;
    width: auto;
    position: absolute;
    top: 100px;
    right: 8px;
    margin: 3px;
    filter: invert(39%) sepia(100%) saturate(2618%) hue-rotate(197deg) brightness(98%) contrast(95%);
}