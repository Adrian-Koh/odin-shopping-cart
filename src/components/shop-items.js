async function shopItems() {
  try {
    const response = await fetch("https://fakestoreapi.com/products");

    if (!response.ok) {
      throw new Error("Error while fetching products");
    }
    const products = await response.json();
    return products;
  } catch (error) {
    console.log(error);
  }
}

export { shopItems };
