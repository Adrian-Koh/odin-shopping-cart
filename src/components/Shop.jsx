import { Link } from "react-router-dom";
import { useEffect, useState } from "react";
import "../styles/Shop.css";
import { shopItems } from "./shop-items";

const Shop = () => {
  const [products, setProducts] = useState([]);

  useEffect(() => {
    const fetchProducts = async () => {
      const result = await shopItems();
      console.log(result);

      setProducts(result);
    };

    fetchProducts();
  }, []);
  return (
    <>
      <h1>Shop</h1>
      <Link to="home">Home</Link>
      <ul>
        {products.map((product) => {
          return (
            <li>
              <img src={product.image} alt={product.title} />
              <h2>{product.title}</h2>
              <p>Category: {product.category}</p>
              <p>Description: {product.description}</p>
            </li>
          );
        })}
      </ul>
    </>
  );
};

export { Shop };
