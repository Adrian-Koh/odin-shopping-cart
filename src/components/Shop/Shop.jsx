import { Link } from "react-router-dom";
import { useEffect, useState } from "react";
import styles from "./Shop.module.css";
import { shopItems } from "../shop-items";
import { CardButtons } from "../CardButtons/CardButtons";
import { Cart } from "../Cart/Cart";

const ALL_CATEGORY = "all products";

const Shop = () => {
  const [products, setProducts] = useState([]);
  const [activeCategory, setActiveCategory] = useState(ALL_CATEGORY);
  const [categories, setCategories] = useState([]);
  const [cartItems, setCartItems] = useState([]);
  const [showShop, setShowShop] = useState(true);

  useEffect(() => {
    const fetchProducts = async () => {
      const result = await shopItems();
      console.log(result);

      setProducts(result);

      const categorySet = new Set(result.map((product) => product.category));
      setCategories([ALL_CATEGORY, ...categorySet]);
    };

    fetchProducts();
  }, []);

  function addCartItem(item) {
    let newCartItems;
    if (cartItems.filter((cartItem) => cartItem.id === item.id).length > 0) {
      newCartItems = cartItems.map((cartItem) => {
        if (cartItem.id === item.id) {
          return { ...cartItem, quantity: cartItem.quantity + item.quantity };
        }
        return cartItem;
      });
    } else {
      newCartItems = [...cartItems, item];
    }
    console.log(newCartItems);
    setCartItems(newCartItems);
  }

  function updateItemFromCart(item) {
    let newCartItems;
    if (cartItems.filter((cartItem) => cartItem.id === item.id).length > 0) {
      newCartItems = cartItems.filter((cartItem) => cartItem.id !== item.id);
      if (item.quantity === 0) {
        setCartItems(newCartItems);
        return;
      }
    }
    newCartItems = [...newCartItems, item];
    setCartItems(newCartItems);
  }

  function handleNavClick(show) {
    setShowShop(show);
  }

  function handleCategoryClick(category) {
    setActiveCategory(category);
    setShowShop(true);
  }

  return (
    <>
      <div className={styles.title}>
        <h1>Shop</h1>
        <nav className={styles.shopNavBar}>
          <img
            src="../../public/shopping.svg"
            alt="shop"
            onClick={() => handleNavClick(true)}
            className={!showShop ? styles.showIcon : styles.hideIcon}
          />
          <img
            src="../../public/cart.svg"
            alt="cart"
            onClick={() => handleNavClick(false)}
            className={showShop ? styles.showIcon : styles.hideIcon}
          />
        </nav>
      </div>
      <div className={styles.container}>
        <aside>
          <h2>Categories</h2>
          <ul>
            {categories.map((category) => {
              return (
                <li
                  key={category}
                  onClick={() => handleCategoryClick(category)}
                >
                  {category}
                </li>
              );
            })}
          </ul>
        </aside>
        <div className={styles.mainContainer}>
          {showShop ? (
            <main>
              {products
                .filter((product) =>
                  activeCategory === ALL_CATEGORY
                    ? product
                    : product.category === activeCategory
                      ? product
                      : null,
                )
                .map((product) => {
                  return (
                    <div className={styles.card} key={product.id}>
                      <img
                        src={product.image}
                        alt={product.title}
                        className={styles.cardImage}
                      />
                      <p className={styles.cardTitle}>{product.title}</p>
                      <p className={styles.price}>
                        ${Number(product.price).toFixed(2)}
                      </p>
                      <img
                        src="../../../public/information.svg"
                        alt="info"
                        className={styles.infoIcon}
                      />
                      <CardButtons
                        updateCartItem={addCartItem}
                        id={product.id}
                        isInCart={false}
                      />
                    </div>
                  );
                })}
            </main>
          ) : (
            <Cart
              products={products}
              cartItems={cartItems}
              updateCartItem={updateItemFromCart}
            />
          )}
        </div>
      </div>
    </>
  );
};

export { Shop };
