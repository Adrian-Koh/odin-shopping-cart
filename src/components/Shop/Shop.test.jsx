import { describe, it, expect, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import { Shop } from "./Shop";
import * as shopItems from "../shop-items";

describe("Shop component", () => {
  it("renders shop", () => {
    const { container } = render(<Shop />);
    expect(container).toMatchSnapshot();
  });

  it("renders shop with correct length of grid items", async () => {
    let id;
    const mockItems = [
      {
        title: "Item A",
        category: "Category 1",
        image: "/dummy/path",
        id: id++,
        price: 10,
        description: "item description of A",
      },
      {
        title: "Item B",
        category: "Category 1",
        image: "/dummy/path",
        id: id++,
        price: 20,
        description: "item description of B",
      },
      {
        title: "Item C",
        category: "Category 2",
        image: "/dummy/path",
        id: id++,
        price: 20,
        description: "item description of C",
      },
    ];
    vi.spyOn(shopItems, "shopItems").mockResolvedValue(mockItems);
    render(<Shop />);

    await screen.findByText("Item A"); // wait for items to complete loading

    expect(document.querySelector("main").children.length).toBe(
      mockItems.length,
    );

    const categoryTwo = screen.getByText("Category 2");
    fireEvent.click(categoryTwo);

    await screen.findByText("Item C");
    expect(document.querySelector("main").children.length).toBe(
      mockItems.filter((item) => item.category === "Category 2").length,
    );
  });
});
