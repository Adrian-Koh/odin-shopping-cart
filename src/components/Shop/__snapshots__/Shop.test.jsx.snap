// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shop component > renders shop 1`] = `
<div>
  <div
    class="_title_6b6c4b"
  >
    <h1>
      The Everything Shop
    </h1>
    <nav
      class="_shopNavBar_6b6c4b"
    >
      <img
        alt="shop"
        class="_hideIcon_6b6c4b"
        src="../../public/shopping.svg"
      />
      <img
        alt="cart"
        class="_showIcon_6b6c4b"
        src="../../public/cart.svg"
      />
    </nav>
  </div>
  <h1>
    Loading...
  </h1>
</div>
`;
